### PR TITLE
Averyyan

### PR DESCRIPTION
**Pull request by averyyan**
add a fun and test to continue UserTask

### CODE REVIEW

File: pkg/bpmn_engine/engine_conditions.go [hunk 1]
```diff
@@ -4,6 +4,8 @@ import (
 	"github.com/nitram509/lib-bpmn-engine/pkg/spec/BPMN20"
 )
 
+/* 流过程条件处理 */
```
Comment (nitram509) [inline pkg/bpmn_engine/engine_conditions.go:7]: Please do comments in English, or leave them out.

File: pkg/bpmn_engine/engine_properties.go [hunk 2]
```diff
@@ -7,10 +7,9 @@ import (
 )
 
 type ProcessInfo struct {
-	BpmnProcessId string // The ID as defined in the BPMN file
-	Version       int32  // A version of the process, default=1, incremented, when another process with the same ID is loaded
-	ProcessKey    int64  // The engines key for this given process with version
-
```
Comment (nitram509) [inline pkg/bpmn_engine/engine_properties.go:13]: why was this line removed?
I propose leaving the former code unchanged.

File: pkg/bpmn_engine/engine_test.go [hunk 3]
```diff
@@ -14,9 +14,10 @@ type CallPath struct {
 
 func (callPath *CallPath) CallPathHandler(job ActivatedJob) {
 	if len(callPath.CallPath) > 0 {
-		callPath.CallPath += ","
+		callPath.CallPath = callPath.CallPath + ","
```
Comment (nitram509) [inline pkg/bpmn_engine/engine_test.go:17]: Why was this code changed?
I propose leaving the former code unchanged.

File: pkg/bpmn_engine/engine_test.go [hunk 4]
```diff
@@ -14,9 +14,10 @@ type CallPath struct {
...
 	}
-	callPath.CallPath += job.GetElementId()
+	callPath.CallPath = callPath.CallPath + job.GetElementId()
+	fmt.Println(job.GetElementId(), "is done", callPath.CallPath)
```
Comment (nitram509) [inline pkg/bpmn_engine/engine_test.go:20]: please, remove this debug-println statement.

File: pkg/bpmn_engine/engine_test.go [hunk 5]
```diff
@@ -28,7 +29,7 @@ func TestRegisterHandlerByTaskIdGetsCalled(t *testing.T) {
 	// setup
 	bpmnEngine := New("name")
 	process, _ := bpmnEngine.LoadFromFile("../../test-cases/simple_task.bpmn")
-	wasCalled := false
+	var wasCalled = false
```
Comment (nitram509) [inline pkg/bpmn_engine/engine_test.go:32]: why was this 'var' introduced?
I propose leaving the former code unchanged.

File: pkg/bpmn_engine/engine_usertasks_test.go [hunk 6]
```diff
@@ -20,3 +21,31 @@ func Test_user_tasks_can_be_handled(t *testing.T) {
...
+
+	userConfirm := false
+	bpmnEngine.AddTaskHandler("user-task", func(job ActivatedJob) {
+		if userConfirm {
+			cp.CallPathHandler(job)
+		}
+	})
+	bpmnEngine.RunOrContinueInstance(instance.instanceKey)
+
+	time.Sleep(3 * time.Second)
```
Comment (nitram509) [inline pkg/bpmn_engine/engine_usertasks_test.go:42]: this time.Sleep() is IMHO not necessary and just delays the overall test execution.
can we remove it?

File: pkg/bpmn_engine/engine.go [hunk 7]
```diff
@@ -184,6 +193,18 @@ func (state *BpmnEngineState) run(instance *ProcessInstanceInfo) (err error) {
 	return err
 }
 
+func (state *BpmnEngineState) findUnCompletedUserTasksForContinuation(process *ProcessInfo, instance *ProcessInstanceInfo) (ret []*BPMN20.TUserTask) {
+	for _, event := range process.definitions.Process.UserTasks {
+		for _, job := range state.jobs {
+			if (job.State == activity.Active || job.State == activity.Ready) && job.ProcessInstanceKey == instance.instanceKey && job.ElementId == event.GetId() {
```
Comment (nitram509) [inline pkg/bpmn_engine/engine.go:199]: Assuming, we allow application developers to exit a handler without calling job.Complete() or job.Fail() ... we only have active jobs to continue.
I do propose:
1. rename the function -> findActiveUserTasksForContinuation
2. just search for job.State == activity.Active

File: pkg/bpmn_engine/engine.go [hunk 8]
```diff
@@ -184,6 +193,18 @@ func (state *BpmnEngineState) run(instance *ProcessInstanceInfo) (err error) {
 	return err
 }
 
+func (state *BpmnEngineState) findUnCompletedUserTasksForContinuation(process *ProcessInfo, instance *ProcessInstanceInfo) (ret []*BPMN20.TUserTask) {
+	for _, event := range process.definitions.Process.UserTasks {
```
Comment (nitram509) [inline pkg/bpmn_engine/engine.go:197]: the local loop variable should rather be named ```userTask``` instead of event

File: pkg/bpmn_engine/engine.go [hunk 9]
```diff
@@ -114,7 +114,16 @@ func (state *BpmnEngineState) run(instance *ProcessInstanceInfo) (err error) {
 		}
 		instance.state = process_instance.ACTIVE
 	case process_instance.ACTIVE:
+		//  find UnCompleted UserTasks from back store to continue
+		userTasks := state.findUnCompletedUserTasksForContinuation(process, instance)
+		for _, ice := range userTasks {
```
Comment (nitram509) [inline pkg/bpmn_engine/engine.go:119]: the local loop variable should rather be named ```userTask``` instead of ice

**Review by nitram509 — CHANGES_REQUESTED** [review]

**Comment by nitram509** [comment]
Hi @averyyan , 
thank you very much for this contribution.
I left some comments/feedback on your PR - where overall I see this as an interesting enhancement.

May I ask, what is your motivation/use case, you try to achieve with this?

Also, I see this change would let service tasks and user tasks be handled differently.
That might be intended, but I would appreciate some documentation (see ./docs folder).

File: pkg/bpmn_engine/engine_usertasks_test.go (hunk 6 above)
Comment (averyyan) [reply pkg/bpmn_engine/engine_usertasks_test.go:42]: yes，I just use it to simulation a action

File: pkg/bpmn_engine/engine.go (hunk 7 above)
Comment (averyyan) [reply pkg/bpmn_engine/engine.go:199]: My earliest version just like this.

**Comment by averyyan** [comment]
> Hi @averyyan , thank you very much for this contribution. I left some comments/feedback on your PR - where overall I see this as an interesting enhancement.
> 
> May I ask, what is your motivation/use case, you try to achieve with this?
> 
> Also, I see this change would let service tasks and user tasks be handled differently. That might be intended, but I would appreciate some documentation (see ./docs folder).

I write a mongodb as back store. try to use db to store Job.

Some times it must rebuild process from db. 

I write a version has break changes to use mongodb.

findUnCompletedUserTasksForContinuation this fun can be use both in UserTask and ServiceTask

File: pkg/bpmn_engine/engine_properties.go (hunk 2 above)
Comment (averyyan) [reply pkg/bpmn_engine/engine_properties.go:13]: I am not use github well. I have a mongodb branch. i think this is a merge mistaken

**Comment by averyyan** [comment]
> Hi @averyyan , thank you very much for this contribution. I left some comments/feedback on your PR - where overall I see this as an interesting enhancement.
> 
> May I ask, what is your motivation/use case, you try to achieve with this?
> 
> Also, I see this change would let service tasks and user tasks be handled differently. That might be intended, but I would appreciate some documentation (see ./docs folder).

I have repair comments you left. Do I need push my code again？

**Comment by nitram509** [comment]
> I have repair comments you left. Do I need push my code again？

Yes, please.
When doing a review on/with Github you can simply push changes on the same branch,
from which you made the pull requests.
Once you've made all your changes and pushed them, just click on the review button (top right in a PR)
or simply ping me a message.

**Comment by nitram509** [comment]
> I am not use github well. I have a mongodb branch. i think this is a merge mistaken

No problem. That can happen to the best developers as well ;)
It's good that you have realized/mentioned it. 👍

**Review by averyyan — COMMENTED** [review]
try
